src/ui/component/userSignIn/view.jsx

```diff
@@ -4,98 +4,132 @@ import { withRouter } from 'react-router';
+    emailToVerify,
+    user,
+    claimingReward,
+    claimedRewards,
+    channels,
+    claimReward,
+    balance,
+    history,
+    location,
+    fetchUser,
+    youtubeChannels,
+    syncEnabled,
+    syncingWallet,
+    getSyncError,
+    hasSynced,
+    fetchingChannels,
+  } = props;
   const { search } = location;
   const urlParams = new URLSearchParams(search);
   const redirect = urlParams.get('redirect');
-  const hasFetchedReward = useFetched(claimingReward);
   const hasVerifiedEmail = user && user.has_verified_email;
   const rewardsApproved = user && user.is_reward_approved;
+  const hasFetchedReward = useFetched(claimingReward);
   const channelCount = channels ? channels.length : 0;
-  const hasFetchedChannels = channels !== undefined;
   const hasClaimedEmailAward = claimedRewards.some(reward => reward.reward_type === REWARDS.TYPE_CONFIRM_EMAIL);
-  const memoizedClaimReward = React.useCallback(() => {
-    claimReward();
-  }, [claimReward]);
+  const hasYoutubeChannels = youtubeChannels && Boolean(youtubeChannels.length);
+  const hasTransferrableYoutubeChannels = hasYoutubeChannels && youtubeChannels.some(channel => channel.transferable);
+  const hasPendingYoutubeTransfer =
+    hasYoutubeChannels && youtubeChannels.some(channel => channel.transfer_state === 'pending_transfer');
+
+  // Complexity warning
+  // We can't just check if we are currently fetching something
+  // We may want to keep a component rendered while something is being fetched, instead of replacing it with the large spinner
+  // The verbose variable names are an attempt to alleviate _some_ of the confusion from handling all edge cases that come from
+  // reward claiming (plus the balance updating after), channel creation, account syncing, and youtube transfer
+  const canHijackSignInFlowWithSpinner = hasVerifiedEmail && !hasClaimedEmailAward && balance === 0 && !getSyncError;
+  const isCurrentlyFetchingSomething = fetchingChannels || claimingReward || syncingWallet;
+  const isWaitingForSomethingToFinish =
+    // If the user has claimed the email award, we need to wait until the balance updates sometime in the future
+    !hasFetchedReward || (hasFetchedReward && balance === 0) || (syncEnabled && !hasSynced);
+
+  // The possible screens for the sign in flow
+  const showEmail = !emailToVerify && !hasVerifiedEmail;
+  const showEmailVerification = emailToVerify && !hasVerifiedEmail;
+  const showUserVerification = hasVerifiedEmail && !rewardsApproved;
+  const showSyncPassword = syncEnabled && getSyncError && !hasSynced;
+  const showChannelCreation =
+    hasVerifiedEmail && balance && balance > DEFAULT_BID_FOR_FIRST_CHANNEL && channelCount === 0 && !hasYoutubeChannels;
+  const showYoutubeTransfer =
+    hasVerifiedEmail && hasYoutubeChannels && (hasTransferrableYoutubeChannels || hasPendingYoutubeTransfer);
+  const showLoadingSpinner =
+    canHijackSignInFlowWithSpinner && (isCurrentlyFetchingSomething || isWaitingForSomethingToFinish);
```

Comment:
I don't think you've done anything wrong here, but this seems quite complicated. I'd be curious if you have any thoughts on to how this complexity could be reduced.

Reply:
I was hoping you would have some thoughts 😛 

Most of the complexity comes from dealing with multiple types of state and api calls. For example, claiming the email verified reward. We can't just check that you claimed it, but that the balance has been updated _after_ just claiming it.